### ISSUE TITLE
Modularization of wireguard/agent

### ISSUE THREAD

**Issue by bimmlerd** [cilium/cilium#25262]
The WireGuard Agent is responsible for managing transparent encryption using WireGuard in cilium, see the [docs](https://docs.cilium.io/en/v1.13/security/network/encryption-wireguard/#encryption-wg).

The code in `pkg/wireguard` is already pretty self-contained. Current `NewAgent()` and `Agent.Init()` methods map onto constructor and start hook, though the private key loading or generation should live in the start hook instead of the constructor. 

A slight wart here is the check against IPSec - WireGuard and IPSec are two incompatible features. We should move feature compatability checks out of the individual cells into some early config validation thing.

## Dependencies

`Init()` shows two dependencies: `IPCache` and the MTU configuration. The former seems easy, it already exists in the object graph. However, we have a cycle in the dependency graph: `Datapath -> wireguard.Agent -> IPCache -> PolicyTrifecta -> Datapath`. The IPCache dependency stems from the need to update WireGuard peers' AllowedIPs when necessary, might need some introduction logic at startup instead of hard depending. 

The MTU dependency is nontrivial too, #23888 tracks the modularization of the MTU handling. Might be possible to split out the specific thing we need from the general MTU modularization, to unblock this.

We also have a dependency on the local node (to check `OptOutNodeEncryption` and to set the pubkey for the node), as well as the `DaemonConfig` for node to node encryption. The local node stuff can be handled once the LocalNodeStore work is in, that's the first three commits of #24866. 

Finally, the agent also depends on `option.Config.StateDir` to be around, since that's where the private key file lives. It's not just the path, but also the fact that the directory exists - that probably wants to be some sort of `BaseEnvironment` cell that establishes some of the common runtime directories.

## Future work

Device manager for netlink setup stuff of wireguard interface?

Ref: #23425
Blocked by #23888

**Comment by bprashanth** [comment]
Thanks for the detailed write up @bimmlerd 

Is it feasible to refactor the daemon so one does NOT require a restart to enable encryption? (eg say the daemon watches the configmap and calls the right functions "on the fly"). 

I am asking because a rolling restart of the agent across a large cluster could take hours.
